### PR TITLE
Revocations

### PR DESCRIPTION
**Pull request by jeffsantarossa**
Issue #6

### CODE REVIEW

**Comment by hjacobs** [comment]
We certainly need to work on the code coverage (to keep it high) :smirk:

File: revoke/revokeprovider.go [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+	// check global revocation
+	if r := crp.cache.Get("GLOBAL"); r != nil && r.(*Revocation).Timestamp > iat {
+		log.Printf("Found GLOBAL revocation")
+		return true
+	}
+
+	// check token revocation
+	th := hashTokenClaim(j.Raw)
+	if r := crp.cache.Get(th); r != nil && r.(*Revocation).Timestamp < iat {
+		log.Printf("Found TOKEN revocation. Hash: %s", th)
```
Comment (hjacobs) [inline revoke/revokeprovider.go]: I guess we should not log the fact that the token was revoked anywhere, but we should have metrics for it (how many requests hit revoked tokens by revocation type) :smile:

File: revoke/revokeprovider.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+			return true
+		}
+	}
+
+	return false
+}
+
+func hashTokenClaim(h string) string {
+
+	//	salt := options.HashingSalt
```
Comment (hjacobs) [inline revoke/revokeprovider.go]: I guess "options.HashingSalt" will move back in?

Comment (jeffsantarossa) [inline revoke/revokeprovider.go]: Good find, @hjacobs!

File: revoke/caching.go [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+		return
+	}
+	c.forceRefresh <- ts
+}
+
+func (c *Cache) Add(rev *Revocation) {
+	var hash string
+	switch rev.Type {
+	case "TOKEN":
+		hash = rev.Data["token_hash"].(string)
```
Comment (hjacobs) [inline revoke/caching.go]: this might panic

File: revoke/revokeprovider.go [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+	// check claim revocation
+	cn := crp.cache.GetClaimNames()
+	if len(cn) == 0 {
+		return false
+	}
+	for _, n := range cn {
+		val, ok := j.Claims[n]
+		// claim might not be present in this JWT!
+		if ok {
```
Comment (hjacobs) [inline revoke/revokeprovider.go]: and test for the "not ok" case (I added the "if" as it was throwing a panic if the JWT does not contain the claim)

File: revoke/revokeprovider.go [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+	crp.cache.Expire()
+
+}
+
+func (crp *CachingRevokeProvider) IsJWTRevoked(j *jwt.Token) bool {
+
+	iat := int(j.Claims["iat"].(float64))
+
+	// check global revocation
+	if r := crp.cache.Get("GLOBAL"); r != nil && r.(*Revocation).Data["issued_before"].(int) > iat {
```
Comment (hjacobs) [inline revoke/revokeprovider.go]: this might panic if the "Revocation" data is not as expected...

File: revoke/job.go [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+import (
+	"time"
+)
+
+type JobFunc func()
+
+func Schedule(interval time.Duration, job JobFunc) {
+	go func() {
+		for {
```
Comment (vetinari) [inline revoke/job.go]: if you want the job to happen in fixed intervals, no matter how long job() takes, use a Ticker:
```go
func job() {
    time.Sleep(500 * time.Millisecond)
}
func main() {
    for _ = range time.Tick(1 * time.Second) {
        fmt.Printf("NOW=%s\n", time.Now())
        job()
    }
}
```

File: revoke/revokeprovider.go [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+
+	"github.com/dgrijalva/jwt-go"
+	"github.com/rcrowley/go-metrics"
+	"github.com/zalando/planb-tokeninfo/breaker"
+	"github.com/zalando/planb-tokeninfo/options"
+)
+
+// always refresh with additional tolerance to make sure we catch all revocations
+// (server clocks might not be synchronized, Cassandra replication might be delayed)
+const refreshToleranceSeconds = 60
```
Comment (vetinari) [inline revoke/revokeprovider.go]: export it, so it can be set from outside this package if required later

File: revoke/caching.go [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+				key := ""
+				for k, rev := range c {
+					if rev.(*Revocation).Timestamp > t {
+						t = rev.(*Revocation).Timestamp
+						key = k
+					}
+				}
+				r.res <- c[key]
+			case r := <-cName:
+				names := make(map[string]int)
```
Comment (vetinari) [inline revoke/caching.go]: why "int"? :)

File: revoke/revokeprovider.go [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+
+	jr := new(jsonRevoke)
+	if err := jr.UnmarshallJSON(body); err != nil {
+		log.Println("Failed to unmarshall revocation data. " + err.Error())
+		return
+	}
+
+	if jr.Meta.RefreshTimestamp != 0 {
+		r := crp.cache.Get("FORCEREFRESH")
+		if r == nil || (r != nil && r.(*Revocation).Timestamp != jr.Meta.RefreshTimestamp) {
```
Comment (vetinari) [inline revoke/revokeprovider.go]: minor: the `r != nil` is not needed (keep parens)

File: revoke/caching.go [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+			log.Println("Error adding revocation to cache: missing claim value_hash.")
+			return
+		}
+		hash = rev.Data["value_hash"].(string)
+	case REVOCATION_TYPE_GLOBAL:
+		hash = REVOCATION_TYPE_GLOBAL
+	case REVOCATION_TYPE_FORCEREFRESH:
+		hash = REVOCATION_TYPE_FORCEREFRESH
+	default:
+		return
```
Comment (vetinari) [inline revoke/caching.go]: drop a note in the log about the unknown rev.Type?

File: revoke/revokeprovider.go [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+
+	if resp.StatusCode != http.StatusOK {
+		log.Printf("Failed to get revocations. Server returned status %s.", resp.Status)
+		return
+	}
+
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+
+	jr := new(jsonRevoke)
```
Comment (vetinari) [inline revoke/revokeprovider.go]: jr := &jsonRevoke{}

File: revoke/revoke.go [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+	Data      map[string]interface{}
+	Timestamp int
+}
+
+type jsonRevoke struct {
+	Meta struct {
+		RefreshFrom      int `json:"REFRESH_FROM"`
+		RefreshTimestamp int `json:"REFRESH_TIMESTAMP"`
+	} `json:"meta"`
+	Revs []jsonRevocation `json:"revocations"`
```
Comment (vetinari) [inline revoke/revoke.go]: []*jsonRevocation?

File: revoke/caching.go [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+
+	return names
+}
+
+func (c *Cache) Expire() {
+	c.expire <- true
+}
+
+func (c *Cache) ForceRefresh(ts int) {
+	if ts <= 0 {
```
Comment (vetinari) [inline revoke/caching.go]: compare "less than oldest valid token"

File: revoke/caching.go [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+				}
+			case r := <-ts:
+				t := 0
+				key := ""
+				for k, rev := range c {
+					if rev.(*Revocation).Timestamp > t {
+						t = rev.(*Revocation).Timestamp
+						key = k
+					}
+				}
```
Comment (vetinari) [inline revoke/caching.go]: if v, ok := c[key]; ok { r.res <- v } else { r.res <- nil }

**Pull request merged**